lib/prefix.c

```diff
@@ -1361,92 +1361,92 @@ char *evpn_es_df_alg2str(uint8_t df_alg, char *buf, int buf_len)
 }
 
 printfrr_ext_autoreg_p("EA", printfrr_ea)
-static ssize_t printfrr_ea(char *buf, size_t bsz, const char *fmt,
-			   int prec, const void *ptr)
+static ssize_t printfrr_ea(struct fbuf *buf, struct printfrr_eargs *ea,
+			   const void *ptr)
 {
 	const struct ethaddr *mac = ptr;
+	char cbuf[ETHER_ADDR_STRLEN];
 
-	if (mac)
-		prefix_mac2str(mac, buf, bsz);
-	else
-		strlcpy(buf, "NULL", bsz);
+	if (!mac)
+		return bputs(buf, "(null)");
 
-	return 2;
+	/* need real length even if buffer is too short */
+	prefix_mac2str(mac, cbuf, sizeof(cbuf));
+	return bputs(buf, cbuf);
 }
 
 printfrr_ext_autoreg_p("IA", printfrr_ia)
-static ssize_t printfrr_ia(char *buf, size_t bsz, const char *fmt,
-			   int prec, const void *ptr)
+static ssize_t printfrr_ia(struct fbuf *buf, struct printfrr_eargs *ea,
+			   const void *ptr)
 {
 	const struct ipaddr *ipa = ptr;
+	char cbuf[INET6_ADDRSTRLEN];
 
-	if (ipa)
-		ipaddr2str(ipa, buf, bsz);
-	else
-		strlcpy(buf, "NULL", bsz);
+	if (!ipa)
+		return bputs(buf, "(null)");
 
-	return 2;
+	ipaddr2str(ipa, cbuf, sizeof(cbuf));
+	return bputs(buf, cbuf);
 }
 
 printfrr_ext_autoreg_p("I4", printfrr_i4)
-static ssize_t printfrr_i4(char *buf, size_t bsz, const char *fmt,
-			   int prec, const void *ptr)
+static ssize_t printfrr_i4(struct fbuf *buf, struct printfrr_eargs *ea,
+			   const void *ptr)
 {
-	if (ptr)
-		inet_ntop(AF_INET, ptr, buf, bsz);
-	else
-		strlcpy(buf, "NULL", bsz);
+	char cbuf[INET_ADDRSTRLEN];
+
+	if (!ptr)
+		return bputs(buf, "(null)");
 
-	return 2;
+	inet_ntop(AF_INET, ptr, cbuf, sizeof(cbuf));
+	return bputs(buf, cbuf);
 }
 
 printfrr_ext_autoreg_p("I6", printfrr_i6)
-static ssize_t printfrr_i6(char *buf, size_t bsz, const char *fmt,
-			   int prec, const void *ptr)
+static ssize_t printfrr_i6(struct fbuf *buf, struct printfrr_eargs *ea,
+			   const void *ptr)
 {
-	if (ptr)
-		inet_ntop(AF_INET6, ptr, buf, bsz);
-	else
-		strlcpy(buf, "NULL", bsz);
+	char cbuf[INET6_ADDRSTRLEN];
+
+	if (!ptr)
+		return bputs(buf, "(null)");
 
-	return 2;
+	inet_ntop(AF_INET6, ptr, cbuf, sizeof(cbuf));
+	return bputs(buf, cbuf);
 }
 
 printfrr_ext_autoreg_p("FX", printfrr_pfx)
-static ssize_t printfrr_pfx(char *buf, size_t bsz, const char *fmt,
-			    int prec, const void *ptr)
+static ssize_t printfrr_pfx(struct fbuf *buf, struct printfrr_eargs *ea,
+			    const void *ptr)
 {
-	if (ptr)
-		prefix2str(ptr, buf, bsz);
-	else
-		strlcpy(buf, "NULL", bsz);
+	char cbuf[PREFIX_STRLEN];
+
+	if (!ptr)
+		return bputs(buf, "(null)");
 
-	return 2;
+	prefix2str(ptr, cbuf, sizeof(cbuf));
```

Comment:
Feels like a change to `prefix2str`, not the printfrr bits?

Reply:
(btw, the `frr-format` plugin will warn if one of the unsupported prefix types is passed in since they're not listed in the `#pragma FRR printfrr_ext`; that said an unsupported type could still be passed in as the generic `struct prefix *`)